yjit/src/codegen.rs

```diff
@@ -2477,12 +2477,12 @@ fn gen_equality_specialized(
             side_exit,
         );
 
-        let equal = asm.new_label("equal");
         let ret = asm.new_label("ret");
 
         // If they are equal by identity, return true
         asm.cmp(a_opnd, b_opnd);
-        asm.je(equal);
+        asm.mov(C_RET_OPND, Qtrue.into());
```

Comment:
Let's not use C_RET_OPND like this, it's pretty hacky because it touches into low-level workings of the backend and potentially could cause problems with register allocation. The solution with pushing to the Ruby stack in two different places is much cleaner IMO.

Reply:
Sure, I wrote the patch after reading Alan's comment before reading yours, but I agree it's too hacky. So I left it as is https://github.com/Shopify/ruby/pull/371/commits/f08dcf3bc720e45251647c899222aa1e44198e46. It's technically not adding a new branch but just extra `jmp` calls, so hopefully there's no significant performance impact. Just a little bit bigger code, I guess.

Reply:
Not sure why the extra jmp is needed. I want to take time to review this more carefully.

Reply:
It's like:

#### Before
```
  cmp a, b
  rax = true
  je ret
  rax = rb_str_equal_internal(a, b)
ret:
  mov stack, rax
```

#### After
```
  cmp a, b
  je equal
  rax = rb_str_equal_internal(a, b)
  mov stack, rax
  jmp ret
equal:
  mov stack, Qtrue
ret:
```

If we don't have the extra jmp, we would write the stack twice on the ccall path.